CRM/Utils/Mail.php

```diff
@@ -185,6 +177,118 @@ public static function send(array &$params): bool {
+        unset($headers['Bcc']);
+      }
+    }
+
+    if (is_object($mailer)) {
+      try {
+        $result = $mailer->send($to, $headers, $message);
+      }
+      catch (Exception $e) {
+        \Civi::log()->error('Mailing error: ' . $e->getMessage());
+        CRM_Core_Session::setStatus(ts('Unable to send email. Please report this message to the site administrator'), ts('Mailing Error'), 'error');
+        return FALSE;
+      }
+      if (is_a($result, 'PEAR_Error')) {
+        $message = self::errorMessage($mailer, $result);
+        // append error message in case multiple calls are being made to
+        // this method in the course of sending a batch of messages.
+        \Civi::log()->error('Mailing error: ' . $message);
+        CRM_Core_Session::setStatus(ts('Unable to send email. Please report this message to the site administrator'), ts('Mailing Error'), 'error');
+        return FALSE;
+      }
+      // CRM-10699
+      CRM_Utils_Hook::postEmailSend($params);
+      return TRUE;
+    }
+    return FALSE;
+  }
+
+  /**
+   * Send a test email using the selected mailer.
+   *
+   * @param Mail $mailer
+   * @param array $params
+   *   Params by reference.
+   *
+   * @return bool
+   *   TRUE if a mail was sent, else FALSE.
+   */
+  public static function sendTest($mailer, array &$params): bool {
+    CRM_Utils_Hook::alterMailParams($params, 'testEmail');
+    $message = $params['text'];
+    $to = $params['toEmail'];
+
+    list($headers, $message) = self::setEmailHeaders($params);
+
+    $from = self::pluckEmailFromHeader($headers['From']);
+
+    try {
+      $mailer->send($to, $headers, $message);
+
+      $testMailStatusMsg = ts('Sending test email') . ':<br />'
+        . ts('From: %1', [1 => $from]) . '<br />'
+        . ts('To: %1', [1 => $to]) . '<br />';
+      if (defined('CIVICRM_MAIL_LOG') && defined('CIVICRM_MAIL_LOG_AND_SEND')) {
+        $testMailStatusMsg .= '<br />' . ts('You have defined CIVICRM_MAIL_LOG_AND_SEND - mail will be logged.') . '<br /><br />';
+      }
+      if (defined('CIVICRM_MAIL_LOG') && !defined('CIVICRM_MAIL_LOG_AND_SEND')) {
+        CRM_Core_Session::setStatus($testMailStatusMsg . ts('You have defined CIVICRM_MAIL_LOG - no mail will be sent.  Your %1 settings have not been tested.', [1 => strtoupper($mailer->_driver)]), ts("Mail not sent"), "warning");
+      }
+      else {
+        CRM_Core_Session::setStatus($testMailStatusMsg . ts('Your %1 settings are correct. A test email has been sent to your email address.', [1 => strtoupper($mailer->_driver)]), ts("Mail Sent"), "success");
+      }
+    }
+    catch (Exception $e) {
+      $result = $e;
+      Civi::log()->error($e->getMessage());
+      $errorMessage = CRM_Utils_Mail::errorMessage($mailer, $result);
+      CRM_Core_Session::setStatus($testMailStatusMsg . ts('Oops. Your %1 settings are incorrect. No test mail has been sent.', [1 => strtoupper($mailer->_driver)]) . $errorMessage, ts("Mail Not Sent"), "error");
```

Comment:
Does the %1 param work for you? I get a php error about passing null to strtoupper(), because `_driver` is a protected var and is returning null. `$mailer->getDriver()` seems to work.

I also get undefined variable $testMailStatusMsg - copy/paste?

Reply:
I've fixed both issues I think.